### PR TITLE
Various improvements to Context API (doc, naming, deadlines, semantic clarity)

### PR DESCRIPTION
**Pull request by louiscryan**
@ejona86 
@kluever
@jhump 
@netdpb

Various naming and documentation clenaups to make the distinction between 'cascading cancellation' and 'propagating context' clearer.

Remove attachAsCloseable and replace with the immediate run / call methods and encourage folks to  use those rather than the attach/detach methods which have sharper edges

Introduce a concrete Deadline type and use it to avoid common pitfalls around specification of time intervals

Still need to write tests for Deadline but worth sharing as we're still iterating on API shape

### CODE REVIEW

File: core/src/main/java/io/grpc/Context.java [hunk 1]
```diff
@@ -94,6 +91,9 @@
  *   }
  * </pre>
  *
+ * <p>Contexts can also be created with a timeout relative to the system nano clock which will
+ * cancellationCause it to automatically cancel at the desired time.
```
Comment (jhump) [inline core/src/main/java/io/grpc/Context.java]: overreaching rename refactor?

File: core/src/main/java/io/grpc/Context.java [hunk 2]
```diff
@@ -73,8 +68,10 @@
 
  * </pre>
  *
- * <p>Context objects will cascade cancellation from their parent and propagate it to their
- * children. You can add a {@link CancellationListener} to a context to be notified when it or
+ *
+ * <p>Contexts are also be used to represent a scoped unit of work. When the unit of work is
```
Comment (jhump) [inline core/src/main/java/io/grpc/Context.java]: s/also be used/also used/

File: core/src/main/java/io/grpc/Deadline.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * Create a deadline that will expire at the specified offset from the current system clock.
+   * @param duration A non-negative duration.
+   * @param units The time unit for the duration.
+   * @return A new deadline.
+   */
+  public static Deadline after(long duration, TimeUnit units) {
+    Preconditions.checkNotNull(units);
+    if (duration > 0) {
+      return new Deadline(System.nanoTime() + units.toNanos(duration), false);
```
Comment (jhump) [inline core/src/main/java/io/grpc/Deadline.java]: Just in case, maybe perform overflow check?

File: core/src/main/java/io/grpc/Context.java [hunk 4]
```diff
@@ -510,38 +526,38 @@ public C call() throws Exception {
 
   /**
    * Wrap an {@link Executor} so that it executes with this context as the {@link #current} context.
-   * It is generally expected that {@link #propagate(Executor)} would be used more commonly than
-   * this method.
+   * It is generally expected that {@link #currentContextExecutor(Executor)} would be used more
```
Comment (adriancole) [inline core/src/main/java/io/grpc/Context.java]: Late comment, but if fixed context is edge case vs current, maybe have an example of where it might be useful?

File: core/src/test/java/io/grpc/ContextTest.java [hunk 5]
```diff
@@ -506,7 +505,7 @@ public Object call() {
   @Test
   public void propagatingExecutor() throws Exception {
```
Comment (adriancole) [inline core/src/test/java/io/grpc/ContextTest.java]: Nit test names need updating

**Comment by adriancole** [comment]
Sfsg!

File: core/src/main/java/io/grpc/Context.java (hunk 1 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: IntelliJ fist waving. Fixed here and elsewhere

File: core/src/main/java/io/grpc/Context.java (hunk 2 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: Done

File: core/src/main/java/io/grpc/Deadline.java (hunk 3 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: Done

File: core/src/main/java/io/grpc/Context.java (hunk 4 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: Added not about sharding work as a potential use-case

File: core/src/test/java/io/grpc/ContextTest.java (hunk 5 above)
Comment (louiscryan) [inline core/src/test/java/io/grpc/ContextTest.java]: Done

**Comment by louiscryan** [comment]
One of the issues I haven't yet addressed here is the potential distinction between 'done' and 'cancelled' in CancellableContext.

File: core/src/main/java/io/grpc/Context.java [hunk 6]
```diff
@@ -261,12 +249,11 @@ public CancellableContext withDeadlineNanoTime(long deadlineNanoTime,
    *   });
    * </pre>
    */
-  public CancellableContext withDeadlineAfter(long duration, TimeUnit unit,
+  public CancellableContext withDeadline(Deadline deadline,
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Context.java]: Could we keep `withDeadlineAfter` as a convenience? I didn't think anyone had a problem with it, and it was the one I was expecting to be used most by far.

File: core/src/main/java/io/grpc/Context.java [hunk 7]
```diff
@@ -566,65 +585,53 @@ private Object lookup(Key<?> key) {
...
         ScheduledExecutorService scheduler) {
       this(parent);
-      scheduledFuture = scheduler.schedule(new Runnable() {
-        @Override
-        public void run() {
-          cancel(new TimeoutException("context timed out"));
-        }
-      }, delayNanos, TimeUnit.NANOSECONDS);
+      Deadline parentDeadline = DEADLINE_KEY.get(parent);
+      if (parentDeadline == null || deadline == deadline.minimum(parentDeadline)) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Context.java]: I'm not sure how much you want to deal with these cases. Note that if `deadline == parentDeadline` this schedules a timer.

File: core/src/main/java/io/grpc/Context.java [hunk 8]
```diff
@@ -566,65 +585,53 @@ private Object lookup(Key<?> key) {
...
       // cancellable context from Context.current()
-      dummy = new Context(this, EMPTY_ENTRIES);
+      uncancellableSurrogate = new Context(this, EMPTY_ENTRIES);
     }
 
     /**
      * Create a cancellable context that has a deadline.
      */
-    private CancellableContext(Context parent, long delayNanos,
+    private CancellableContext(Context parent, Deadline deadline,
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Context.java]: `deadline` needs to be set as a Context value.

File: core/src/main/java/io/grpc/Deadline.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+/**
+ * An absolute deadline in system time.
+ */
+public class Deadline {
+
+  private static final AtomicIntegerFieldUpdater<Deadline> expiredUpdater =
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: Why not just make expired volatile? You don't use compareAndSet.

(Note that really, you don't even need it volatile, since if it is out of date you will just do an additional time check before setting it to `1` again.)

File: core/src/main/java/io/grpc/Deadline.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  private Deadline(long baseInstant, long offset, boolean baseInstantAlreadyExpired) {
+    if (offset > 0 && Long.MAX_VALUE - offset < baseInstant) {
+      deadlineNanos = Long.MAX_VALUE;
+      expired = 0;
+    } else if (offset < 0 && Long.MIN_VALUE - offset > baseInstant) {
+      deadlineNanos = Long.MIN_VALUE;
+      expired = 1;
+    } else {
+      deadlineNanos = baseInstant + offset;
+      expired = baseInstantAlreadyExpired && offset <= 0 ? 1 : 0;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: Checking `offset` here seems redundant. Either check it in `after()` or check it here, but not both.

File: core/src/main/java/io/grpc/Deadline.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    return new Deadline(deadlineNanos,  units.toNanos(offset), expiredUpdater.get(this) == 1);
+  }
+
+  /**
+   * Schedule a task to be run when the deadline expires.
+   * @param task to run on expiration
+   * @param scheduler used to execute the task
+   * @return {@link ScheduledFuture} which can be used to cancel execution of the task
+   */
+  public ScheduledFuture<?> runOnExpiration(Runnable task, ScheduledExecutorService scheduler) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: Possibly in lieu of this method, it seems we need a way to convert to a timeout, like `long timeRemaining(TimeUnit)`. Whether it returns negative numbers or not is up for discussion. (I _think_ I'm in favor of the negatives.)

File: core/src/main/java/io/grpc/Deadline.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    Preconditions.checkNotNull(task, "task");
+    Preconditions.checkNotNull(scheduler, "scheduler");
+    return scheduler.schedule(task, Math.max(0, deadlineNanos - System.nanoTime()),
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: Note that the max() is unnecessary since negative delays are fine:
> Zero and negative delays (but not periods) are also allowed in schedule methods, and are treated as requests for immediate execution.

File: core/src/main/java/io/grpc/Context.java [hunk 13]
```diff
@@ -49,13 +46,11 @@
...
- * A context propagation mechanism which carries deadlines, cancellation signals,
- * and other scoped values across API boundaries and between threads. Examples of functionality
- * propagated via context include:
+ * A context propagation mechanism which can carry scoped-values across API boundaries and between
+ * threads. Examples of state propagated via context include:
  * <ul>
- *   <li>Deadlines for a local operation or remote call.</li>
  *   <li>Security principals and credentials.</li>
- *   <li>Local and distributed tracing context.</li>
+ *   <li>Local and distributed tracing information.</li>
```
Comment (jhspaybar) [inline core/src/main/java/io/grpc/Context.java]: I'd opened an issue recently about these comments, and my understanding was that tracing information will not propagate "across API boundaries" but that the data which crosses API boundaries is limited to deadlines and cancellations.  Instead, if a user wanted to propagate tracing information(or arbitrary metadata), they would need to write a custom interceptor to read the context and add these values in a header.  I still think this can be made more clear here how users are expected to propagate arbitrary metadata outside of cancellation and deadlines.

File: core/src/main/java/io/grpc/Context.java (hunk 6 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: Sure. Done

File: core/src/main/java/io/grpc/Context.java (hunk 7 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: Fixed check to be more precise. Seems worth doing

File: core/src/main/java/io/grpc/Context.java (hunk 8 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: Done

File: core/src/main/java/io/grpc/Deadline.java (hunk 9 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: Agreed about just using volatile (changed)

I guess it depends how much we care about the memory barrier and the likelihood of calls coming from multiple threads vs. the cost of the additional call to nanoTime.

Looking at
http://shipilev.net/blog/2014/nanotrusting-nanotime/
it seems like the cost of the volatile write ~2ns vs the system call at aout 25ns. Volatile reads are much cheaper, particularly if they are uncontended

File: core/src/main/java/io/grpc/Deadline.java (hunk 10 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: Done

File: core/src/main/java/io/grpc/Deadline.java (hunk 11 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: This method has a tighter contract

Whats the use case for this other method? We can have both if we need to add it later

File: core/src/main/java/io/grpc/Deadline.java (hunk 12 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: Done

File: core/src/main/java/io/grpc/Context.java (hunk 13 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Context.java]: Suggestions for how to document the intent are welcome. I used a hypothetical 'credential' in the example in the class heading as that's a very common case. I'm not sure that documenting the interceptor pattern here is the best idea.

Most of the code to do what you want is in the Contexts class. i.e. in your interceptor you would do....

class TracingInterceptor implements ServerInteceptor {
  <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
      MethodDescriptor<ReqT, RespT> method,
      ServerCall<RespT> call,
      Metadata headers,
      ServerCallHandler<ReqT, RespT> next) {
      Trace trace = header.get(TRACE_KEY); // TRACE_KEY is a Metadata.Key
      return Contexts.interceptCall(
            Context.current().with(CONTEXT_TRACE_KEY, trace), // CONTEXT_TRACE_KEY is a Context.Key
            method, call, headers, next);
}

File: core/src/main/java/io/grpc/Deadline.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+   * @return A new deadline.
+   */
+  public static Deadline after(long duration, TimeUnit units) {
+    Preconditions.checkNotNull(units);
+    return new Deadline(System.nanoTime(), units.toNanos(duration), true);
+  }
+
+  private final long deadlineNanos;
+
+  @SuppressWarnings("unused")
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: It is used :)

File: core/src/main/java/io/grpc/Deadline.java (hunk 11 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: We know gRPC needs the timeout for sending on the wire, so we can't actually use Deadline for gRPC with the current API. I wouldn't find it strange if the user wanted to do some sort of wait() or poll() with a timeout, or you may need to use an API that uses timeouts instead of deadlines. Converting to/from deadline/timeout seems a core use case.

Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: Agreed. Added

File: core/src/test/java/io/grpc/ContextTest.java [hunk 15]
```diff
@@ -555,99 +554,108 @@ public void typicalCancellableTryCatchFinallyHandling() throws Exception {
...
+  }
 
-    try (Closeable c = base.attachAsCloseable()) {
-      // Do something
-      throw new IllegalStateException("Argh");
-    } catch (IllegalStateException ise) {
-      // Don't capture exception
+  @Test
+  public void earlierParentDeadlineTakesPrecedenceOverLaterChildDeadline() throws Exception {
+    final Deadline sooner = Deadline.after(2, TimeUnit.SECONDS);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/ContextTest.java]: Wouldn't a deadline after 0 s, 1 ms, or 100 ms be equally effective without taking as long?

Comment (louiscryan) [inline core/src/test/java/io/grpc/ContextTest.java]: Sure, made 100ms

File: core/src/main/java/io/grpc/Deadline.java (hunk 14 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/Deadline.java]: Done

**Comment by ejona86** [comment]
@louiscryan LGTM. There's a checkstyle failure

**Comment by louiscryan** [comment]
Resolved checkstyle issues. 

Ping 

@kluever
@jhump 
@netdpb
@adriancole 

hopefully this is headed in the right direction. API is marked as Experimental so we can still iterate on it after this

**Comment by jhump** [comment]
LGTM

**Comment by adriancole** [comment]
Yep I like it!
